chpl/chpl-service/src/main/java/gov/healthit/chpl/questionableactivity/ListingQuestionableActivityProvider.java

```diff
@@ -663,6 +569,119 @@ private boolean hasICS(CertifiedProductSearchDetails listing) {
+        QuestionableActivityListingDTO activity = null;
+        if (ff4j.check(FeatureList.EFFECTIVE_RULE_DATE)) {
+            // If ICS=0 and they attested to D2, D3, or D10
+            if (!newListing.getIcs().getInherits()) {
+                List<CertificationResult> matchingCertResults = newListing.getCertificationResults().stream()
+                        .filter(cr -> cr.isSuccess() &&
+                                (cr.getCriterion().getId().equals(d2Criterion.getId())
+                                        || cr.getCriterion().getId().equals(d3Criterion.getId())
+                                        || cr.getCriterion().getId().equals(d10Criterion.getId())))
+                        .collect(Collectors.toList());
+
+                if (matchingCertResults.size() > 0) {
+                    String criteriaNumbers = matchingCertResults.stream()
+                            .map(cr -> cr.getCriterion().getNumber())
+                            .collect(Collectors.joining(", "));
+                    activity = new QuestionableActivityListingDTO();
+                    activity.setAfter(criteriaNumbers);
+                }
+            }
+        }
+        return activity;
+    }
+
+    public QuestionableActivityListingDTO checkNonCuresAuditCriteriaOnEdit(CertifiedProductSearchDetails origListing,
+            CertifiedProductSearchDetails newListing) {
+        QuestionableActivityListingDTO activity = null;
+        if (ff4j.check(FeatureList.EFFECTIVE_RULE_DATE)) {
+            List<String> matchingCriteriaNumbers = new ArrayList<String>();
+            if (hasCriteriaChangedToAttestedTo(d2Criterion, newListing, origListing)) {
+                matchingCriteriaNumbers.add(d2Criterion.getNumber());
+            }
+            if (hasCriteriaChangedToAttestedTo(d3Criterion, newListing, origListing)) {
+                matchingCriteriaNumbers.add(d3Criterion.getNumber());
+            }
+            if (hasCriteriaChangedToAttestedTo(d10Criterion, newListing, origListing)) {
+                matchingCriteriaNumbers.add(d10Criterion.getNumber());
+            }
+
+            if (matchingCriteriaNumbers.size() > 0) {
+                String criteriaNumbers = matchingCriteriaNumbers.stream()
+                        .collect(Collectors.joining(", "));
+                activity = new QuestionableActivityListingDTO();
+                activity.setAfter(criteriaNumbers);
+            }
+        }
+        return activity;
+    }
+
+    public QuestionableActivityListingDTO checkNonCuresAuditCriteriaAndAddedIcsOnEdit(CertifiedProductSearchDetails origListing,
+            CertifiedProductSearchDetails newListing) {
+        QuestionableActivityListingDTO activity = null;
+        if (ff4j.check(FeatureList.EFFECTIVE_RULE_DATE)) {
+            if (!origListing.getIcs().getInherits() && newListing.getIcs().getInherits()) { // ICS has been added
```

Comment:
Could getIcs() be null? I am thinking specifically if this code gets called when editing a legacy 2014 listing? I'm not sure it does get called but just bringing it up.